piet-wgsl/src/shaders/preprocess.rs

```diff
@@ -71,29 +87,53 @@ pub fn preprocess(
                             item.active = !item.active;
                         }
                     }
+                    let remainder = directive_start[directive_len..].trim();
+                    if !remainder.is_empty() {
+                        eprintln!("#else directives don't take an argument. `{remainder}` will not be in output (line {line_number})");
+                    }
+                    continue 'outer;
                 }
                 "endif" => {
                     if let None = stack.pop() {
                         eprintln!("Mismatched endif (line {line_number})");
                     }
+                    continue 'outer;
                 }
                 "import" => {
-                    let import = imports.get(arg);
+                    output.push_str(&line[..hash_index]);
+                    let directive_end = &directive_start[directive_len..];
+                    let import_name_start = if let Some(import_name_start) =
+                        directive_end.find(|c: char| !c.is_whitespace())
+                    {
+                        import_name_start
+                    } else {
+                        eprintln!("#import needs a non_whitespace argument (line {line_number})");
+                        continue 'outer;
+                    };
+                    let import_name_start = &directive_end[import_name_start..];
+                    let import_name_end_index = import_name_start
+                        .find(|c| !(matches!(c, '_') || c.is_alphanumeric()))
```

Comment:
Any special reason `matches!` rather than just `c == '_'`? Looking at the code I'm surprised this compiles without `c` having an explicit type, maybe matches is forcing type equality while PartialEq would be looser.

Reply:
I think orginally there were several characters I wanted to allow there (e.g. `-`, I think)

It does also make the types work out, although obviously it's reasonable to be consistent.